bootstrap/app/addon-manager.go

```diff
@@ -0,0 +1,306 @@
+	}
+	if amle := env("ADDON_MANAGER_LEADER_ELECTION"); amle != "" {
+		b, err := strconv.ParseBool(amle)
+		if err != nil {
+			return nil, fmt.Errorf("error converting ADDON_MANAGER_LEADER_ELECTION value %q to bool: %v", amle, err)
+		}
+		am.leaderElection = b
+	}
+	if testIntervalSec := env("TEST_ADDON_CHECK_INTERVAL_SEC"); testIntervalSec != "" {
+		secs, err := strconv.ParseInt(testIntervalSec, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("error converting TEST_ADDON_CHECK_INTERVAL_SEC value %q to int64: %v", testIntervalSec, err)
+		}
+		am.checkInterval = time.Duration(secs) * time.Second
+	}
+	am.pruneWhitelistFlags = makePruneWhitelistFlags(env)
+	return am, nil
+}
+
+// makePruneWhitelistFlags generates kubectl prune-whitelist flags from provided environment variables.
+func makePruneWhitelistFlags(env func(key string) string) []string {
+	pw := defaultKubectlPruneWhitelistResources
+	if pwo := env("KUBECTL_PRUNE_WHITELIST_OVERRIDE"); pwo != "" {
+		pw = strings.Fields(pwo)
+	}
+	pwf := appendPruneWhitelistFlags(nil, pw)
+	pwf = appendPruneWhitelistFlags(pwf, strings.Fields(env("KUBECTL_EXTRA_PRUNE_WHITELIST")))
+	klog.Infof("== Generated kubectl prune whitelist flags: %s ==", strings.Join(pwf, " "))
+	return pwf
+}
+
+// appendPruneWhitelistFlags generates kubectl prune-whitelist flags from provided resource list.
+func appendPruneWhitelistFlags(pwf, resources []string) []string {
+	for _, resource := range resources {
+		pwf = append(pwf, "--prune-whitelist", resource)
+	}
+	return pwf
+}
+
+func (m *addonManager) Run() error {
+	m.waitForSystemServiceAccount()
+	if err := m.ensureDefaultAdmissionControlsObjects(); err != nil {
+		if os.IsNotExist(err) {
+			klog.Warningf("Admissions control object directory not found: %v", err)
+		} else {
+			klog.Errorf("Could not create admissions control objects: %v", err)
+		}
+	}
+	ticker := time.NewTicker(m.checkInterval)
+	for ; true; <-ticker.C { // Run immediately, then after every tick.
+		if !m.leaderElection || m.isLeader() {
+			m.ensureAddons()
+			m.reconcileAddons()
+		} else {
+			klog.Info("== Another addon-manager is acting as leader. Sleeping... ==")
+		}
+	}
+	return nil
+}
+
+func (m *addonManager) waitForSystemServiceAccount() {
+	var out, errOut bytes.Buffer
+	for {
+		err := m.kubectl(&out, &errOut, m.kubectlOpts, "get", "-n", systemNamespace, "serviceaccount", "default", "-o", "go-template={{with index .secrets 0}}{{.name}}{{end}}")
+		if errStr := errOut.String(); errStr != "" {
+			klog.Warning(errStr)
+		}
+		if err == nil && out.String() != "" {
+			return
+		}
+		out.Reset()
+		errOut.Reset()
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// Create admission_control objects if defined before any other addon services. If the limits
+// are defined in a namespace other than default, we should still create the limits for the
+// default namespace.
+func (m *addonManager) ensureDefaultAdmissionControlsObjects() error {
+	var out, errOut bytes.Buffer
```

Comment:
Any reason to define these buffers outside the retry loop / should you reset it in between retries?

Reply:
just to reduce mallocs, though it's probably not a big deal. probably do want to clear them after each loop, though